source/common/network/io_socket_handle_impl.cc

```diff
@@ -227,7 +229,25 @@ Api::IoCallUint64Result IoSocketHandleImpl::sendmsg(const Buffer::RawSlice* slic
 }
 
 Address::InstanceConstSharedPtr
-maybeGetDstAddressFromHeader(const cmsghdr& cmsg, uint32_t self_port, os_fd_t fd, bool v6only) {
+IoSocketHandleImpl::getOrCreateEnvoyAddressInstance(sockaddr_storage ss, socklen_t ss_len) {
+  if (recent_received_addresses_ == nullptr) {
+    return Address::addressFromSockAddrOrDie(ss, ss_len, fd_,
+                                             socket_v6only_ || !udp_read_normalize_addresses_);
+  }
+  quic::QuicSocketAddress quic_address(ss);
+  auto it = recent_received_addresses_->Lookup(quic_address);
+  if (it != recent_received_addresses_->end()) {
+    return *it->second;
+  }
+  Address::InstanceConstSharedPtr new_address = Address::addressFromSockAddrOrDie(
+      ss, ss_len, fd_, socket_v6only_ || !udp_read_normalize_addresses_);
+  recent_received_addresses_->Insert(
+      quic_address, std::make_unique<Address::InstanceConstSharedPtr>(new_address));
```

Comment:
Why is there a `unique_ptr` to a `shared_ptr`? Looking at the quiche lru it looks like it requires a `unique_ptr` to the key type. Would it make sense to have our own implementation of an lru cache (a very simple data structure) that allows the value type to be a shared_ptr to avoid an additional dereference for every address we have to lookup?

Reply:
The unique_ptr to shared_ptr is just because the quiche LRU cache interface requires that. But the additional dereferencing doesn't stand up in the cpu cost at all.